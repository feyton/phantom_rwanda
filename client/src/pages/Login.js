import { axiosBase as axios } from '@utils/Api.js';
import React, { useEffect, useState } from 'react';
import { useForm } from 'react-hook-form';
import { useDispatch } from 'react-redux';
import { Link, useLocation, useNavigate } from 'react-router-dom';
import { toast } from 'react-toastify';
import busMapImg from '../assets/busMap.png';
import { ButtonLoading } from '../components/Button.js';
import { loginUser } from '../redux/reducers/authReducer.js';

const Login = () => {
  const navigate = useNavigate();
  const location = useLocation();
  const dispatch = useDispatch();
  const [err, setErr] = useState('');
  const [attempts, setAttempts] = useState(0);
  const [loading, setLoading] = useState(false);
  const [passwordShown, setPasswordShown] = useState(false);
  let from = location?.state || '/dashboard/main';
  const {
    register,
    handleSubmit,
    formState: { errors }
  } = useForm();

  const togglePassword = () => {
    setPasswordShown(!passwordShown);
  };
  /* istanbul ignore next */
  const onValid = async (data) => {
    setLoading(true);
    try {
      const response = await axios.post('/accounts/login', data);
      dispatch(loginUser(response.data.data));
      toast(`Welcome ${response.data.data.first_name}`, {
        type: 'info'
      });
      navigate(from);
    } catch (error) {
      setErr(error?.response?.data?.data?.message || error.message);
    } finally {
      setLoading(false);
    }
  };

  const onErrors = (errors) => {
    if (errors.password && !errors.email) {
      setAttempts(attempts + 1);
    }
  };
  useEffect(() => {
    if (attempts > 3 && attempts < 5) {
      setErr('Wrong attempts of more than 3 times.');
    } else if (attempts > 5) {
      setErr('Try again in 5 minutes');
    }
  }, [attempts]);

  return (
    <div className=" z-0 h-full">
      <main className="min-h-[80vh] flex flex-col  md:flex-row items-center w-full justify-center shadow-md relative h-full pb-52 md:pb-1 md:pt-1 pt-10">
        <section className="max-w-sm mt-[-60px] md:flex md:items-center md:justify-center md:mx-auto xl:ml-64">
          <img
            src={busMapImg}
            alt="vector image of a bus on map"
            className="w-full object-cover "
          />
        </section>
        <section className="flex flex-col  py-3 px-4 w-72 md:w-fit md:min-w-[300px] shadow-main absolute bg-white top-28 md:top-0 md:mx-auto md:relative rounded-lg xl:mr-64">
          <form
            id="loginForm"
            className="flex flex-col w-full px-4"
            onSubmit={handleSubmit(onValid, onErrors)}
          >
            <h2 className="text-2xl mb-5 text-center font-extrabold">Login</h2>

            <p className="font-bold text-sm font-raleway">
              {from !== '/dashboard/main' && (
                <div>
                  <p>Login is required to access:</p>
                  <p className="font-bold text-green-600 cursor-pointer">
                    {from}
                  </p>
                </div>
              )}
            </p>
            <p className="font-bold capitalize text-red text-sm py-1">{err}</p>
            <label
              htmlFor="email"
              className="capitalize mb-1.5 font-semibold text-sm required"
            >
              Email:
            </label>
            <input
              type="email"
              id="email"
              name="email"
              placeholder="Enter email"
              {...register('email', {
                required: 'Email is required',
                pattern: {
                  value:
                    /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/,
                  message: 'Enter a valid email'
                }
              })}
              className="rounded-md mb-1 p-1 border-2 w-full outline-none focus:border-gray-500"
            />
            <p id="email-errors" className="text-red-600 text-sm pb-3">
              {errors?.email && errors.email.message}
            </p>
            <label
              htmlFor="email"
              className="capitalize mb-2 font-semibold required"
            >
              Password:
            </label>
            <input
              type={passwordShown ? 'text' : 'password'}
              id="password"
              name="password"
              placeholder="Enter password"
              {...register('password', { required: 'Password is required' })}
              className="rounded-md mb-2 p-1 border-2 w-full outline-none focus:border-gray-500"
            />
            <p className="text-red-600 text-sm">
              {errors?.password && errors.password.message}
            </p>
            <div>
              <input onChange={(e)=>togglePassword(e)} type="checkbox" name="" id="show" />
              <label className='font-raleway font-bold mx-1 text-sm' htmlFor="show">Show password</label>
            </div>

           
            <Link to="/accounts/reset-password">
              <h4 className="text-primary text-right mb-5 text-sm font-bold">
                Forgot password?
              </h4>
            </Link>

            {attempts > 5 ? (
              <button
                disabled
                className="bg-red-500 px-5 py-2 rounded-md text-white w-fit mx-auto hover:bg-red-900 transition-all hover:transition-all"
              >
                Unlock in 5 minutes
              </button>
            ) : (
              <>
                {loading ? (
                  <div className="mx-auto">
                    <ButtonLoading name="sending..." />
                  </div>
                ) : (
                  <button
                    id="login-btn"
                    className="bg-primary px-5 py-2 rounded-md text-white w-fit mx-auto hover:bg-hover transition-all hover:transition-all"
                  >
                    Login
                  </button>
                )}
              </>
            )}
          </form>
        </section>
      </main>
    </div>
  );
};

export default Login;
