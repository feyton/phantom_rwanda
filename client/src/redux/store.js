import { configureStore } from '@reduxjs/toolkit';
import { authReducer } from './reducers/authReducer.js';
import { busMovementReducer } from './reducers/busMovementReducer.js';
import { counterReducer } from './reducers/counterReducer.js';
import { dashboardReducer } from './reducers/dashboardReducer.js';
import { managementReducer } from './reducers/managementReducer.js';
import { profileReducer } from './reducers/profileReducer.js';
import { simulateReducer } from './reducers/simulateReducer.js';
import { socketReducer } from './reducers/socketReducer.js';
import { updateProfileReducer } from './reducers/updateProfileReducer.js';

export const store = configureStore({
  reducer: {
    counter: counterReducer,
    profileUpdates: updateProfileReducer,
    profileInfo: profileReducer,
    auth: authReducer,
    management: managementReducer,
    dashboard: dashboardReducer,
    socket: socketReducer,
    simulate: simulateReducer,
    movement: busMovementReducer
  }
});
