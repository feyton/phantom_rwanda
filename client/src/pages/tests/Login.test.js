import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { mount } from 'enzyme';
import mockAxios from 'jest-mock-axios';
import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter, MemoryRouter } from 'react-router-dom';
import renderer, { act } from 'react-test-renderer';
import { store } from '../../redux/store.js';
import Login from '../Login.js';

afterEach(() => {
  mockAxios.reset();
});
const navigateMock = () => {
  return jest.fn();
};
const wrapper = renderer.create(
  <Provider store={store}>
    <BrowserRouter>
      <Login />
    </BrowserRouter>
  </Provider>
);
const mockedUsedNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUsedNavigate
}));
const mockedUsedLocation = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => mockedUsedLocation
}));

jest.mock('react-hook-form', () => ({
  ...jest.requireActual('react-hook-form'),
  register: jest.fn(),
  handleSubmit: jest.fn()
}));
const mockUseLocationValue = {
  pathname: '/testroute',
  search: '',
  hash: '',
  state: null
};
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: jest.fn().mockImplementation(() => {
    return mockUseLocationValue;
  })
}));

describe('Render the errors', () => {
  beforeEach(() => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Login navigate={navigateMock} onValid={jest.fn()} />
        </BrowserRouter>
      </Provider>
    );
  });
  it('Render email error', async () => {
    fireEvent.submit(screen.getByTestId('loginForm'));
    expect(await screen.findAllByText(/Email is required/i)).toHaveLength(1);
  });
  it('Should not display errors when the form is valid', async () => {
    fireEvent.input(screen.getByTestId('email'), {
      target: {
        value: 'test@gmail.com'
      }
    });
    fireEvent.input(screen.getByTestId('password'), {
      target: {
        value: 'Test@123'
      }
    });
    fireEvent.submit(screen.getByTestId('loginForm'));
    mockAxios.mockResolvedValue({
      data: []
    });
    await waitFor(() => {
      expect(screen.queryAllByText(/Enter a valid email/i)).toHaveLength(0);
    });
  });
});

describe('Login snapshot test', () => {
  it('should render Login', () => {
    const elem = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <Login />
          </BrowserRouter>
        </Provider>
      )
      .toJSON();
    expect(elem).toMatchSnapshot();
  });
  it('test the onValid function', () => {
    const onValid = jest.fn(wrapper.getInstance(), 'onValid');
    expect(onValid).toMatchSnapshot();
  });
});

describe('Attempts', () => {
  it('Should return an attempts error', async () => {
    await act(async () => {
      render(
        <Provider store={store}>
          <BrowserRouter>
            <Login navigate={navigateMock} />
          </BrowserRouter>
        </Provider>
      );
      fireEvent.input(screen.getByTestId('email'), {
        target: {
          value: 'test@gmail.com'
        }
      });
      fireEvent.submit(screen.getByTestId('loginForm'));
      expect(await screen.findAllByText(/Password is required/i)).toHaveLength(
        1
      );
      fireEvent.submit(screen.getByTestId('loginForm'));
      expect(await screen.findAllByText(/Password is required/i)).toHaveLength(
        1
      );
      fireEvent.submit(screen.getByTestId('loginForm'));
      expect(await screen.findAllByText(/Password is required/i)).toHaveLength(
        1
      );
      fireEvent.submit(screen.getByTestId('loginForm'));
      expect(
        await screen.findAllByText(/Wrong attempts of more than 3 times./i)
      ).toHaveLength(1);
      fireEvent.submit(screen.getByTestId('loginForm'));
      expect(
        await screen.findAllByText(/Wrong attempts of more than 3 times./i)
      ).toHaveLength(1);
      fireEvent.submit(screen.getByTestId('loginForm'));
      expect(await screen.findAllByText(/Unlock in 5 minutes/i)).toHaveLength(
        1
      );
      fireEvent.submit(screen.getByTestId('loginForm'));
      expect(
        await screen.findAllByText(/Try again in 5 minutes/i)
      ).toHaveLength(1);
    });
  });
});

describe('Login page functionality tests', () => {
  const handleSubmitMock = jest.fn();
  const wrapper = mount(
    <Provider store={store}>
      <MemoryRouter>
        <Login handleSubmit={handleSubmitMock()} />)
      </MemoryRouter>
    </Provider>
  );
  const simulateOnChangeInput = (wrapper, inputSelector, newValue) => {
    const input = wrapper.find(inputSelector);
    input.simulate('change', {
      target: { value: newValue }
    });
  };
  const loginBtn = wrapper.find('#login-btn');
  it('calls handleSubmit function on form submit', async () => {
    await act(() => {
      loginBtn.simulate('click');
      expect(handleSubmitMock).toBeCalledTimes(1);
    });
  });
});
