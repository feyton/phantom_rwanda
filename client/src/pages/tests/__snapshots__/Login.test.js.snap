// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login snapshot test should render Login 1`] = `
<div
  className=" z-0 h-full"
>
  <main
    className="min-h-[80vh] flex flex-col  md:flex-row items-center w-full justify-center shadow-md relative h-full pb-52 md:pb-1 md:pt-1 pt-10"
  >
    <section
      className="max-w-sm mt-[-60px] md:flex md:items-center md:justify-center md:mx-auto xl:ml-64"
    >
      <img
        alt="vector image of a bus on map"
        className="w-full object-cover "
        src="busMap.png"
      />
    </section>
    <section
      className="flex flex-col  py-3 px-4 w-72 md:w-fit md:min-w-[300px] shadow-main absolute bg-white top-28 md:top-0 md:mx-auto md:relative rounded-lg xl:mr-64"
    >
      <form
        className="flex flex-col w-full px-4"
        id="loginForm"
        onSubmit={[Function]}
      >
        <h2
          className="text-2xl mb-5 text-center font-extrabold"
        >
          Login
        </h2>
        <p
          className="font-bold text-sm font-raleway"
        />
        <p
          className="font-bold capitalize text-red text-sm py-1"
        >
          
        </p>
        <label
          className="capitalize mb-1.5 font-semibold text-sm required"
          htmlFor="email"
        >
          Email:
        </label>
        <input
          className="rounded-md mb-1 p-1 border-2 w-full outline-none focus:border-gray-500"
          id="email"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Enter email"
          type="email"
        />
        <p
          className="text-red-600 text-sm pb-3"
          id="email-errors"
        />
        <label
          className="capitalize mb-2 font-semibold required"
          htmlFor="email"
        >
          Password:
        </label>
        <input
          className="rounded-md mb-2 p-1 border-2 w-full outline-none focus:border-gray-500"
          id="password"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Enter password"
          type="password"
        />
        <p
          className="text-red-600 text-sm"
        />
        <div>
          <input
            id="show"
            name=""
            onChange={[Function]}
            type="checkbox"
          />
          <label
            className="font-raleway font-bold mx-1 text-sm"
            htmlFor="show"
          >
            Show password
          </label>
        </div>
        <a
          href="/accounts/reset-password"
          onClick={[Function]}
        >
          <h4
            className="text-primary text-right mb-5 text-sm font-bold"
          >
            Forgot password?
          </h4>
        </a>
        <button
          className="bg-primary px-5 py-2 rounded-md text-white w-fit mx-auto hover:bg-hover transition-all hover:transition-all"
          id="login-btn"
        >
          Login
        </button>
      </form>
    </section>
  </main>
</div>
`;

exports[`Login snapshot test test the onValid function 1`] = `[MockFunction]`;
